File: Server/src/main/java/com/dotori/dotori/post/entity/Post.java
package com.dotori.dotori.post.entity;

import com.dotori.dotori.auth.entity.Auth;
import jakarta.persistence.*;
import lombok.*;
import org.springframework.data.annotation.CreatedDate;
import org.springframework.data.annotation.LastModifiedDate;
import org.springframework.data.jpa.domain.support.AuditingEntityListener;

import java.time.LocalDateTime;
import java.util.ArrayList;
import java.util.HashSet;
import java.util.List;
import java.util.Set;

@Entity
@Getter
@Builder
@Setter
@ToString(exclude = "tags")
@AllArgsConstructor
@NoArgsConstructor
@EntityListeners(value = {AuditingEntityListener.class})
public class Post{

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    @Column(name = "pid")
    private Long pid;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "auth_id")
    private Auth auth;

    @Column(name = "nickName")
    private String nickName;

    @Column(nullable = false,length = 500)
    private String content;

    @CreatedDate
    @Builder.Default
    @Column(name = "regDate", updatable = false)
    private LocalDateTime regDate = LocalDateTime.now();

    @LastModifiedDate
    @Column(name = "modDate")
    private LocalDateTime modDate;

    @OneToMany(mappedBy = "post", cascade = CascadeType.ALL, orphanRemoval = true)
    @Builder.Default
    private List<PostThumbnail> thumbnails = new ArrayList<>();

    @ManyToMany
    @JoinTable(
            name = "post_tag",
            joinColumns = @JoinColumn(name = "post_id"),
            inverseJoinColumns = @JoinColumn(name = "tag_id")
    )
    @Builder.Default
    private List<Tag> tags = new ArrayList<>();

    // cascade 추가하여 연관된 엔티티 먼저 삭제
    @OneToMany(mappedBy = "post", cascade = CascadeType.ALL, orphanRemoval = true)
    @Builder.Default
    private List<Comment> comments = new ArrayList<>();

    @Builder.Default
    @ManyToMany(cascade = {CascadeType.PERSIST, CascadeType.MERGE})
    @JoinTable(
            name = "post_mentions",
            joinColumns = @JoinColumn(name = "post_id"),
            inverseJoinColumns = @JoinColumn(name = "auth_id")
    )
    private Set<Auth> mentionedUsers = new HashSet<>();

    public void changePost(String content, LocalDateTime modDate, List<PostThumbnail> thumbnails) {
        this.content = content;
        this.modDate = modDate;
        this.thumbnails = thumbnails;
    }

    public void changePost(String content) {
        this.content = content;
    }

    public void addThumbnail(PostThumbnail postThumbnail) {
        this.thumbnails.add(postThumbnail);
        postThumbnail.setPost(this);
    }

    public void addMention(Auth user) {
        this.mentionedUsers.add(user);
    }

    public void removeMention(Auth user) {
        this.mentionedUsers.remove(user);
    }

    public Set<Auth> getMentionedUsers() {
        return this.mentionedUsers;
    }


}

File: Client-Next/src/post/models/post.js
// Post 클래스: 게시글의 데이터 구조 정의 
export class Post {
  constructor(data) {
    this.pid = data.pid; // 게시글 ID
    this.content = data.content; // 게시글 내용
    this.aid = data.aid; // 작성자 ID (변경)
    this.nickName = data.nickName; // 작성자 닉네임
    this.profileImage = data.profileImage; // 작성자 프로필 이미지
    this.regDate = new Date(data.regDate); // 게시글 등록 날짜
    this.modDate = data.modDate ? new Date(data.modDate) : null; // 게시글 수정 날짜
    this.thumbnails = data.thumbnails; // 게시글 썸네일 이미지 배열
    this.tags = data.tags; // 게시글 태그 배열
    this.toriBoxCount = data.toriBoxCount; // 좋아요 수
    this.commentCount = data.commentCount; // 댓글 수
    this.liked = data.liked; // 현재 사용자의 좋아요 여부
    this.bookmarked = data.bookmarked; // 현재 사용자의 북마크 여부
    this.bookmarkCount = data.bookmarkCount; // 북마크 수
  }
}